inference-engine/src/cldnn_engine/cldnn_infer_request.cpp

```diff
@@ -454,6 +454,160 @@ CLDNNInferRequest::CLDNNInferRequest(InputsDataMap networkInputs, OutputsDataMap
+        streamID = streamExecutor->GetStreamId();
+        int numGraphs = streamGraphs.size();
+        streamID = streamID % numGraphs;
+    }
+    m_graph = streamGraphs[streamID];
+
+    if (m_graph->GetMaxDynamicBatchSize() > 1) {
+        preprocess_dynamic();
+        return;
+    }
+    execDataPreprocessing(_inputs, true);  // "true" stands for serial preprocessing in case of OpenMP
+}
+
+void CLDNNInferRequest::enqueue() {
+    if (m_graph->GetMaxDynamicBatchSize() > 1) {
+        enqueue_dynamic();
+        return;
+    }
+
+    // set input and output memory from request blob maps
+    // into the network object primitives
+    std::vector<cldnn::event::ptr> dependencies;
+    for (auto& item : _inputs) {
+        std::string inputName = item.first;
+        Blob::Ptr& inputBlob = item.second;
+
+        auto nv12_ptr = inputBlob->as<NV12Blob>();
+        auto batched_ptr = inputBlob->as<BatchedBlob>();
+        bool is_batched = batched_ptr != nullptr;
+        bool is_nv12 = nv12_ptr != nullptr;
+
+        if (is_nv12 || is_batched) {
+            int num_blobs = is_batched ? batched_ptr->size() : 1;
+            int expected_batch = is_batched
+                ? _networkInputs.at(inputName)->getTensorDesc().getDims()[0]
+                : 1;
+            for (auto i = 0; i < expected_batch; i++) {
+                std::string y_name = inputName + "_Y" + std::to_string(i);
+                std::string uv_name = inputName + "_UV" + std::to_string(i);
+                if (is_batched) {
+                    int idx = i < num_blobs ? i : num_blobs - 1;
+                    nv12_ptr = getNV12BlobOrException(batched_ptr, idx);
+                }
+                prepare_input(y_name, nv12_ptr->y(), dependencies);
+                prepare_input(uv_name, nv12_ptr->uv(), dependencies);
+            }
+        } else {
+            // regular blob
+            prepare_input(inputName, inputBlob, dependencies);
+        }
+    }
+
+    for (auto& item : _outputs) {
+        std::string outputName = item.first;
+        Blob::Ptr& outputBlob = item.second;
+        prepare_output(outputName, outputBlob);
+    }
+
+    internal_outputs.clear();
+    internal_outputs = m_graph->GetNetwork()->execute(dependencies);
+}
+
+void CLDNNInferRequest::wait() {
+    if (m_graph->GetMaxDynamicBatchSize() > 1) {
+        wait_dynamic();
+        return;
+    }
+
+    if (internal_outputs.empty()) {
+        IE_THROW() << "Inference was not started!\n";
+    }
+
+    // wait for completion & collect outputs as requested by the model
+    for (auto& no : _networkOutputs) {
+        Blob::Ptr bptr = _outputs[no.first];
+        std::string outputID = outputsMap.at(no.first);
+        auto outputMemory = internal_outputs.at(outputID).get_memory();
+
+        // mapping remote blobs not needed -
+        // let the user take care of them explicitly
+        if (!bptr->is<gpu::ClBlob>()) {
+            copy_output_data(outputMemory, bptr);
```

Comment:
Yes, there is an ability to use callbacks. But it is not obvious how this exact call would benefit from this usage.
This line is executed only in case when SetBlob() with user-defined host buffer was called, so it needs to get result buffer to the host before engaging any further pipeline steps anyway.

Reply:
copy_output_data will be called from a single thread? Serialization can degrade perfomance

Reply:
Sorry, but I didn't get completely what exactly do you mean by "called from a single thread".

There can be as many parallel pipeline instances as many ExecutableNetworks or GPU streams have been created.
So, execution/copying for the different streams is run in parallel. But for all requests belonging to same stream, it is serialized.

If you mean that calling this function for several streams in parallel can create weak CPU oversubscription, then the following should be taken into consideration:
1) Since copying is called only for host memory buffers that have been allocated and set by user, there is no way to avoid performing a possibly parallel copy.
2) If user avoids using SetBlob() with host buffers, this function won't be called at all
3) GPU tasks are typically serialized by the driver (except some dGPU cases), and usually copying stages from two GPU pipelines don't overlap.

So, such oversubscription should generally rarely happen, if possible at all.
Anyway, if you gave suggestion how to avoid parallel copy for any case, please go on and describe it in more detail.